docs/operational_guide/bootstrapping.md

```diff
@@ -0,0 +1,112 @@
+
+Generally speaking, we recommend that operators do not modify the bootstrappers configuration, but in the rare case that you to, this document is designed to help you understand the implications of doing so.
+
+M3DB currently supports 5 different bootstrappers:
+
+1. `filesystem`
+2. `commitlog`
+3. `peers`
+4. `uninitialized_topology`
+5. `noop_all`
+
+When the bootstrapping process begins, M3DB nodes need to determine two things:
+
+1. What shards they should bootstrap, which can be determined from the cluster placement.
+2. What time-ranges they need to bootstrap those shards for, which can be determined from the namespace retention.
+
+For example, imagine a M3DB node that is responsible for shards 1, 5, 13, and 25 according to the cluster placement. In addition, it has a single namespace called "metrics" with a retention starting 48 hours ago and ending at the current time. When the M3DB node is started, the node will determine that it needs to bootstrap shards 1, 5, 13, and 25 for the time range starting at the current time and ending 48 hours ago. In order to obtain all this data, it will run the configured bootstrappers in the specified order. Every bootstrapper will notify the bootstrapping process of which shard/ranges it was able to bootstrap and the bootstrapping process will continue working its way through the list of bootstrappers until all the shards/ranges required have been marked as fulfilled. Otherwise the M3DB node will fail to start.
+
+### Filesystem Bootstrapper
+
+The `filesystem` bootstrapper's responsibility is to determine which immutable [fileset files](../m3db/architecture/storage.md) exist on disk, and if so, mark them as fulfilled. The `filesystem` bootstrapper achieves this by scanning M3DB's directory structure and determining which fileset files already exist on disk. Unlike the other bootstrappers, the `filesystem` bootstrapper does not need to load any data into memory, it simply verifies the checksums of the data on disk and other components of the M3DB node will handle reading (and caching) the data dynamically once it begins to serve reads.
+
+### Commitlog Bootstrapper
+
+The `commitlog` bootstrapper's responsibility is to read the `commitlog` and snapshot (compacted commitlogs) files on disk and recover any data that has not yet been written out as an immutable fileset file. Unlike the `filesystem` bootstrapper, the commit log bootstrapper cannot simply check which files are on disk in order to determine if it can satisfy a bootstrap request. Instead, the `commitlog` bootstrapper determines whether it can satisfy a bootstrap request using a simple heuristic.
```

Comment:
hm should this be "`commitlog` and `snapshot` files"

Reply:
it should be all un-block quoted. Fixed